### DISCUSSION TITLE
Is it possible to have an array of a specific class, and can I still manipulate each instance of that class, while it's in that array?

### DISCUSSION THREAD

**Discussion by GrimyKimu** [HaxeFlixel/flixel#2400] [Q&A] [answered]
Okay, I'm still getting the ropes of it both Haxe itself and HaxeFlixel, so things are still a bit confusing:

Can I create an array like the following?
`var testArray:Array<testClass>;`
In this case, "testClass" will be extending FlxSprite, and has an int variable called "id".

Then using that array, can I add instances of that class into it using this in a function?
```haxe
var id:Int = testArray.length();
testArray.push(new testClass(id));

```
And even if I _can_, am I still able to remove, kill, or manipulate those instances from the current state and array?
Would I be able to do something like this:
```haxe
for (i in 0...testArray.length())
{
     testArray[i].testFunction(); //does this even do anything?
     trace(testArray[i].id); //can this work?
     testArray[i].alpha = 100; //will this change the alpha on an individual instance of testClass?
     testArray[i].kill(); //does this kill only an individual instance of testClass?
}
```
Can I manipulate each individual instance of "testClass" in these sorts of ways during runtime?

Forgive me if my syntax/writing is total garbage, I'm still learning. Have some mercy.

The overarching plan is to have multiple copies of the same exact sprite created and placed into the world, but each one is still unique enough that it can be deleted/manipulated individually or in mass. Thanks in advance if you can help~!

**Comment by RichardBray** [comment, marked as answer]
Yes, this is 100% possible. I've done this with a [Map in Haxe ](https://haxe.org/manual/std-Map.html)but I'm sure it can be done with an array.

Let's say you have three sprites, (SpriteOne, SpriteTwo, SpriteTree), you have them all in a folder called 'spritesLove', and you want to loop over them. You could create a constant like this:
```hx
class Constants {
        ...
        public static final spriteNames:Map<String, Class<FlxSprite>> = [
		"Sprite1" => spritesLove.SpriteOne, 
		"Sprite2" => spritesLove.SpriteTwo,
		"Sprite3" => spritesLove.SpriteThree,		
	];
}
```

The strings on the left of the map can have whatever name you want. I personally use a map for save data reasons, so that I just save the string, 'Sprite2', and grab the value like so `final spriteIWant = spriteNames['Sprite2']`.


```hx
  for (sprite in Constants.spriteNames) {
      final spriteClass = Type.createInstance(sprite, []); // you can put class arguments in the empty array
      // Then all the stuff you want to do should work
      spriteClass.testFunction();
      trace(spriteClass.id);
      spriteClass.alpha = 100;
      spriteClass.kill();   
  }
```

**Reply by GrimyKimu** [reply]
Oof, maps huh, can't wait to figure how those work. (_don't worry, I'll figure it out on my own_)
Thanks for the help!

**Reply by Gama11** [reply]
@RichardBray Storing `Class<T>` in a map and using it with `Type.createInstance()` is arguably a bit of an anti-pattern, Haxe has better solutions: https://haxe.org/blog/codingtips-new/

**Reply by RichardBray** [reply]
Wow that's actually much better thanks for sharing @Gama11 
I have an example of an improved method here if anyone is interested:

https://try.haxe.org/#0F2b8FeE

I haven't written Haxe in a while so I'm open to refactoring improvements :)
